sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -287,3 +287,101 @@ case class ArrayContains(left: Expression, right: Expression)
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def children: Seq[Expression] = Seq(x, start, length)
+
+  override def nullSafeEval(xVal: Any, startVal: Any, lengthVal: Any): Any = {
+    val startInt = startVal.asInstanceOf[Int]
+    val lengthInt = lengthVal.asInstanceOf[Int]
+    val arr = xVal.asInstanceOf[ArrayData]
+    val startIndex = if (startInt == 0) {
+      throw new RuntimeException(
+        s"Unexpected value for start in function $prettyName:  SQL array indices start at 1.")
+    } else if (startInt < 0) {
+      startInt + arr.numElements()
+    } else {
+      startInt - 1
+    }
+    if (lengthInt < 0) {
+      throw new RuntimeException(s"Unexpected value for length in function $prettyName: " +
+        s"length must be greater than or equal to 0.")
+    }
+    // this can happen if start is negative and its absolute value is greater than the
+    // number of elements in the array
+    if (startIndex < 0) {
+      return new GenericArrayData(Array.empty[AnyRef])
+    }
+    val elementType = x.dataType.asInstanceOf[ArrayType].elementType
+    val data = arr.toArray[AnyRef](elementType)
+    new GenericArrayData(data.slice(startIndex, startIndex + lengthInt))
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val elementType = x.dataType.asInstanceOf[ArrayType].elementType
+    nullSafeCodeGen(ctx, ev, (x, start, length) => {
+      val arrayClass = classOf[GenericArrayData].getName
+      val values = ctx.freshName("values")
+      val i = ctx.freshName("i")
+      val startIdx = ctx.freshName("startIdx")
+      val resLength = ctx.freshName("resLength")
+      val defaultIntValue = CodeGenerator.defaultValue(CodeGenerator.JAVA_INT, false)
+      s"""
+         |${CodeGenerator.JAVA_INT} $startIdx = $defaultIntValue;
+         |${CodeGenerator.JAVA_INT} $resLength = $defaultIntValue;
+         |if ($start == 0) {
+         |  throw new RuntimeException("Unexpected value for start in function $prettyName: "
+         |    + "SQL array indices start at 1.");
+         |} else if ($start < 0) {
+         |  $startIdx = $start + $x.numElements();
+         |} else {
+         |  // arrays in SQL are 1-based instead of 0-based
+         |  $startIdx = $start - 1;
+         |}
+         |if ($length < 0) {
+         |  throw new RuntimeException("Unexpected value for length in function $prettyName: "
+         |    + "length must be greater than or equal to 0.");
+         |} else if ($length > $x.numElements() - $startIdx) {
+         |  $resLength = $x.numElements() - $startIdx;
+         |} else {
+         |  $resLength = $length;
+         |}
+         |Object[] $values;
+         |if ($startIdx < 0) {
+         |  $values = new Object[0];
+         |} else {
+         |  $values = new Object[$resLength];
+         |  for (int $i = 0; $i < $resLength; $i ++) {
+         |    $values[$i] = ${CodeGenerator.getValue(x, elementType, s"$i + $startIdx")};
```

Comment:
May this assignment cause performance degradation due to boxing if array element type is primitive (e.g. `float`)?

Reply:
I though about that too, but I am not sure there is a better solution: this approach is used both in `CreateArray` and `GenerateSafeProjection`. And there is a TODO for specialized versions of `GenericArrayData` able to deal with primitive types without boxing.

Probably we can try and fix this TODO in another PR/JIRA. What do you think?

Reply:
I see. If we postpone specialization, is it necessary to generate Java code for now? The generated code seems to do the same thing in `nullSafeEval`. WDYT?

Reply:
I think it can be helpful. Moreover, this is the way also `CreateArray` and `GenerateSafeProjection` work, so for coherency I think this is the right thing to do. What do you think?

Reply:
For the future, I agree that this is the right way to generate Java code since we can avoid boxing.

On the other hand, you are proposing to postpone specialization. In `eval` and generated code, `GenericArrayData` is generated by using `Object[]`.
I may misunderstand `for coherency` since I may not find the target of the coherency in the thread.

Reply:
My target of coherency was the `CreateArray` operator and the code generated in `GenerateSafeProjection`.

Reply:
I might miss something, but seems like `CreateArray` is using different ways to codegen for primitive arrays and the others, and I guess `GenerateSafeProjection` is using `Object[]` on purpose to create `GenericArrayData` to be "safe" (avoid using `UnsafeXxx`).
I think we should modify this codegen to avoid boxing. WDYT?

Btw, we need to null check here for an array of primitive type contains null anyway?

Reply:
You are right, I am not sure why I missed it...maybe I checked outdated code. Sorry, I am fixing it, thanks.